### DISCUSSION TITLE
Introducing Scheduled Channel (a.k.a Pre-recorded Live or File Live)

### DISCUSSION THREAD

**Discussion by getroot** [OvenMediaLabs/OvenMediaEngine#1461] [Announcement]
A feature that allows you to create a live channel by scheduling pre-recorded files has been added to OvenMediaEngine. Other services or software call this Pre-recorded Live or File Live, but OvenMediaEngine plans to expand the function to organize live channels as a source, so we named it Scheduled Channel. 

I'd appreciate your feedback if you have a moment. Thank you for your support! This is currently in beta version, and we plan to officially update it once we integrate live sources and complete the implementation for FallbackProgram in the future.

This feature can be used after version 0.16.1 or with the master branch.

## Getting Started

To use this feature, activate Schedule Provider as follows.

```
<VirtualHost>
    <Applications>
        <Providers>
            <Schedule>
                <MediaRootDir>/opt/ovenmediaengine/media</MediaRootDir>
                <ScheduleFilesDir>/opt/ovenmediaengine/media</ScheduleFilesDir>
            </Schedule>
            ...
```

`MediaRootDir`
 Root path where media files are located

`ScheduleFileDir`
 Root path where the schedule file is located

## Schedule Files

Scheduled Channel creates/updates/deletes streams by creating/editing/deleting files with the .sch extension in the ScheduleFileDir path. Schedule files (.sch) use the following XML format. When a `<Stream Name>.sch` file is created in ScheduleFileDir, OvenMediaEngine analyzes the file and creates a Schedule Channel with `<Stream Name>`. If the contents of `<Stream Name>.sch` are changed, the Schedule Channel is updated, and if the file is deleted, the stream is deleted.

```
<?xml version="1.0" encoding="UTF-8"?>
<Schedule>
    <Stream>
        <Name>tv1</Name> <!-- optional, using filename without ext -->
        <BypassTranscoder>false</BypassTranscoder>
        <VideoTrack>true</VideoTrack>
        <AudioTrack>true</AudioTrack>
    </Stream>
    
    <FallbackProgram> <!-- Not yet supported -->
        <Item url="file://sample.mp4" start="0" duration="60000"/>
    </FallbackProgram>

    <Program name="1" scheduled="2023-09-27T13:21:15.123+09:00" repeat="true">
        <Item url="stream://default/app/stream1" duration="60000"/>
    </Program>
    <Program name="2" scheduled="2022-03-14T15:10:0.0+09:00" repeat="true">
        <Item url="file://sample.mp4" start="0" duration="60000"/>
        <Item url="stream://default/app/stream1" duration="60000"/> <!-- Not yet supported -->
        <Item url="file://sample.mp4" start="60000" duration="120000"/>
    </Program>
</Schedule>
```
`Stream (required)`    
 This is the stream information that the Channel needs to create.

`Stream.Name (optional)`    
 It's the stream's name. This is a reference value extracted from the file name for usage. It's recommended to set it same for consistency, although it's for reference purposes.

`Stream.BypassTranscoder (optional, default: false)`
 Set to true if transcoding is not desired.

`Stream.VideoTrack (optional, default: true)`
Determines whether to use the video track. If VideoTrack is set to true and there's no video track in the Item, an error will occur.

`Stream.AudioTrack (optional, default: true)`
Determines whether to use the audio track. If AudioTrack is set to true and there's no audio track in the Item, an error will occur.

`FallbackProgram (optional)`    
Defines the default Item to be broadcasted in the absence of a scheduled Program at the current time or in case of errors in the Item.

`Program (optional)`    
Schedules a program. The `name` is an optional reference value. If not set, a random name will be assigned. Set the start time in ISO8601 format in the `scheduled` attribute. Decide whether to repeat the `Items` when its playback ends.

`Program.Item (optional)`
Configures the media source to broadcast.

The `url` points to the location of the media source. If it starts with 'file://', it refers to a file within the MediaRootDir directory. If it starts with 'stream://', it refers to another stream within the same OvenMediaEngine.

For 'file' cases, the `start` attribute can be set in milliseconds to indicate where in the file playback should start.
`duration` indicates the playback time of that item in milliseconds. After the duration ends, it moves to the next item.
Both 'start' and 'duration' are optional. If not set, `start` defaults to 0, and `duration` defaults to the file's duration; if not specified, the media file will be played until its full duration.

## REST API

Schedule Channel provides the following APIs.

> **GET /v1/vhosts/{vhost}/apps/{app}/scheduledChannels**
    
Request a list of all Scheduled Channels.

> **POST /v1/vhosts/{vhost}/apps/{app}/scheduledChannels**
    
Create a Scheduled channel.

`Request Body Example`
``` 
    {
        "stream": {
            "name": "tv1",
            "bypassTranscoder": false,
            "videoTrack": true,
            "audioTrack": true
        },
        "programs": [
            {
                "name": "1",
                "scheduled": "2023-09-26T13:21:15.123+09:00",
                "repeat": true,
                "items": [
                    {
                        "url": "file://sample.mp4",
                        "start": 0,
                        "duration": 60000
                    },
                    {
                        "url": "stream://default/app/stream1",
                        "duration": 60000
                    }
                ]
            },
            {
                "name": "2",
                "scheduled": "2023-09-27T13:21:15.123+09:00",
                "repeat": true,
                "items": [
                    {
                        "url": "file://sample.mp4",
                        "start": 0,
                        "duration": 60000
                    },
                    {
                        "url": "stream://default/app/stream1",
                        "duration": 60000
                    },
                    {
                        "url": "file://sample.mp4",
                        "start": 60000,
                        "duration": 120000
                    }
                ]
            }
        ]
    }
```

> **PATCH /v1/vhosts/{vhost}/apps/{app}/scheduledChannels/{channel}**

Update the schedule. `<Stream>` cannot be PATCHed.

`Request Body Example`
```
    {
        "programs": [
            {
                "name": "1",
                "scheduled": "2023-09-26T13:21:15.123+09:00",
                "repeat": true,
                "items": [
                    {
                        "url": "file://sample.mp4",
                        "start": 0,
                        "duration": 60000
                    },
                    {
                        "url": "stream://default/app/stream1",
                        "duration": 60000
                    }
                ]
            },
            {
                "name": "2",
                "scheduled": "2023-09-27T13:21:15.123+09:00",
                "repeat": true,
                "items": [
                    {
                        "url": "file://sample.mp4",
                        "start": 0,
                        "duration": 60000
                    },
                    {
                        "url": "stream://default/app/stream1",
                        "duration": 60000
                    },
                    {
                        "url": "file://sample.mp4",
                        "start": 60000,
                        "duration": 120000
                    }
                ]
            }
        ]
    }
```

> **GET /v1/vhosts/{vhost}/apps/{app}/scheduledChannels/{channel}**

Obtain Scheduled Channel information and stat. It also provides information about the currently playing program and item.

> **DELETE /v1/vhosts/{vhost}/apps/{app}/scheduledChannels/{channel}**

Delete the Scheduled Channel.

**Comment by bchah** [comment]
Thank you @getroot for this new and powerful feature! I tried setting up a basic file for playback as specified in the documentation and it works 😀 Transcoding will be important because most offline files are not encoded optimally for streaming - an issue that shows for sure if you bypass the transcoder. I think it would be good to expand the Conditional transcoding options to include BFrame and Keyframe interval options - i.e. transcode only if BFrames are detected or keyframe interval is above N seconds. Then one application could handle both live and scheduled streams without doing unneeded transcoding on the live content. I will test further as soon as possible!

**Reply by getroot** [reply]
Thank you for your feedback.

In most cases, it is recommended to encode Scheduled Channels to keep the output profile constant. The reason I created the `<BypassTranscoder>false</BypassTranscoder>` option is for cases where pre-prepared files are all already encoded with the same profile. If this is not the case, please encode without fail. Most players cannot handle dynamic changes in resolution or encoding options.

In conclusion, when there are multiple items in a Scheduled Channel program, it is almost impossible to encode some items and not encode others using a conditional transcoder. This makes the internal pipeline very complex and very unlikely to play properly. Scheduled Channels must be encoded in most cases.

**Reply by bchah** [reply]
Thank you, that makes sense!

**Comment by bchah** [comment]
I notice this error in the log if there is no duration of 1000 or more specified for a Program Item:

`W [DQAppTicker:3451] ScheduledProvider | scheduled_application.cpp:257  | Failed to update schedule (Could not create schedule): /files/scheduledfiles//helloWorld.sch - Failed to find duration attribute, duration is required`

That is OK but your documentation suggests that omitting the duration will stream the entire file. Perhaps that and/or specifying "0" as the duration.

**Reply by getroot** [reply]
This is not the intended behavior. I'm going to fix this.

**Comment by bchah** [comment]
If you remove or rename a single file that is scheduled, or there are no playable files in the program, the stream gets 'stuck' in a live state and the OME log repeats this every second:

```
[2023-11-29 14:46:10.107] E [Scheduled:4739] ScheduledChannel Provider | scheduled_stream.cpp:471  | #default#live/helloWorld: Failed to open /files/scheduledmedia/test.mov item. error (-2, No such file or directory)
[2023-11-29 14:44:21.539] E [Scheduled:4739] ScheduledChannel Provider | scheduled_stream.cpp:471  | #default#live/helloWorld: Failed to open /files/scheduledmedia/test.mov item. error (-2, No such file or directory)
[2023-11-29 14:44:21.540] E [Scheduled:4739] ScheduledChannel Provider | scheduled_stream.cpp:257  | Scheduled Channel : #default#live/helloWorld: Failed to prepare file playback. Try to play next item
[2023-11-29 14:44:21.541] E [Scheduled:4739] ScheduledChannel Provider | scheduled_stream.cpp:235  | Scheduled Channel #default#live/helloWorld: Too many errors. Sleep for 1000 milliseconds
```

Perhaps if there are no valid program items, the stream should stop?

**Reply by bchah** [reply]
Another interesting symptom in this category. If I let the stream hang for a while with no source file available, then replace the source file, the stream goes "wild" 🤠 It appears to be processing at the maximum possible speed trying to catch up to the elapsed duration of the playback attempt:

https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/bf4f1b07-ed84-4f35-8359-ce2547bc21d2

... and the log output at the time:

```
[2023-11-29 15:55:21.683] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 3613, threshold: 500, peak: 4050
[2023-11-29 15:55:22.460] W [AW-LLHLS0:6061] FMP4 Packager | fmp4_storage.cpp:341  | LLHLS stream (#default#live/helloWorld) / track (0) - a longer-than-expected (5500.0 | expected : 4000) segment has created. This may be due to very long keyframe intervals.
[2023-11-29 15:55:26.715] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 3141, threshold: 500, peak: 4050
[2023-11-29 15:55:31.732] W [InboundWorker:6056] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 2714, threshold: 500, peak: 4050
[2023-11-29 15:55:36.760] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 2346, threshold: 500, peak: 4050
[2023-11-29 15:55:41.786] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 1851, threshold: 500, peak: 4050
[2023-11-29 15:55:46.832] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 1522, threshold: 500, peak: 4050
[2023-11-29 15:55:51.857] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 1124, threshold: 500, peak: 4050
[2023-11-29 15:55:56.877] W [Dech264:6529] ManagedQueue | managed_queue.h:313  | [217] mngq:v=#default#live:s=helloWorld:p=trs:n=decoder_h264_0 size has exceeded the threshold: queue: 725, threshold: 500, peak: 4050
[2023-11-29 15:56:08.310] W [AW-LLHLS0:6061] FMP4 Packager | fmp4_storage.cpp:341  | LLHLS stream (#default#live/helloWorld) / track (0) - a longer-than-expected (6000.0 | expected : 4000) segment has created. This may be due to very long keyframe intervals.
[2023-11-29 15:57:04.108] W [AW-LLHLS0:6061] FMP4 Packager | fmp4_storage.cpp:341  | LLHLS stream (#default#live/helloWorld) / track (0) - a longer-than-expected (6000.0 | expected : 4000) segment has created. This may be due to very long keyframe intervals.
```

The transcoder settings are:

```
<Video>
<Name>original_video</Name>
<Codec>h264</Codec>
<Width>1920</Width>
<Height>1080</Height>
<Bitrate>4000000</Bitrate>
<Framerate>30.0</Framerate>
<KeyFrameInterval>60</KeyFrameInterval>
<BFrames>0</BFrames>
<Preset>slower</Preset>
<ThreadCount>4</ThreadCount>
</Video>
```

**Reply by getroot** [reply]
This is one policy I have thought deeply about and is the behavior I intend to take. This problem is solved when the operator recognizes the mistake and updates the schedule.

- There are live items in the Schedule, but there may or may not be actual live input. To prepare for these situations, I am developing a fallback item function.
- When there is a file item in the schedule, there cannot be a case where the file does not exist. This is because the only case where the file does not exist is when the operator sets it incorrectly. But we'll make it possible to use a fallback item nonetheless.

How to handle the case where the fallback item file does not exist? It's the operator's fault.

**Reply by getroot** [reply]
About playing too fast:
I haven't yet accounted for the case where the file isn't there, so the timer keeps running. I think this could be improved a little further.

**Reply by bchah** [reply]
Thank you, your policy definitely makes sense when tied with the fallback feature!

**Comment by bchah** [comment]
Is there more information available on the supported codecs, containers and parameters for scheduled playback? It seems to have trouble decoding certain clips, and doesn't seem to support advanced codecs like HEVC in my initial testing.

Here is a sample H264 file that for some reason does not want to stream smoothly over Scheduled (with the transcoder enabled as you recommended). Does it stream smoothly on your side?

https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/7b56d517-7431-4497-8d5f-ec24b6ff1a8d

And an HEVC file that should stream fine into OME but the video track is not recognized by the Scheduled provider: 

https://github.com/AirenSoft/OvenMediaEngine/assets/33604674/2bd22b3e-62d4-4f0a-9bee-e18729edd3a6

**Reply by getroot** [reply]
Thank you for your hard work testing. It's a great contribution.
I will be testing these samples soon.

**Reply by getroot** [reply]
When I tested it, test.mov played smoothly. Playing hevc.mov had a bug, so I fixed it and committed it.

**Reply by getroot** [reply]
https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/494953b9-50d7-4398-b2d6-23a2b78650c0

**Comment by getroot** [comment]
I added live input (stream://) and FallbackProgram functions.

Live input is available in the following formats:

`stream://vhost_name/app_name/stream_name`

This will tap (mirroring) data from other streams and output it again from the current Scheduled Channel.

`FallbackProgram `is a program that switches automatically when there is no program scheduled at the current time or an error occurs in an item. If the program is updated at the current time or the item returns to normal, it will fail back to the original program. Both files and live can be used for items in FallbackProgram. However, it is recommended to use a stable file.

This function is a scheduling channel, but it can be used for applications such as creating a permanent stream as follows.

```
<?xml version="1.0"?>
<Schedule>
        <Stream>
                <Name>stream</Name>
                <BypassTranscoder>false</BypassTranscoder>
                <VideoTrack>true</VideoTrack>
                <AudioTrack>true</AudioTrack>
        </Stream>
        <FallbackProgram>
                <Item url="file://hevc.mov"/>
                <Item url="file://avc.mov"/>
        </FallbackProgram>

        <Program name="origin" scheduled="2000-01-01T20:57:00.000+09" repeat="true">
                <Item url="stream://default/app/input" duration="-1" />
        </Program>
</Schedule>
```

This channel normally plays default/app/input, but when live input is stopped, it plays the file in FallbackProgram. This will last forever until the .sch file is deleted. One trick was to set the origin program's schedule time to 2000 so that this stream would play unconditionally.

Then, please give us a lot of feedback. This has quite complex logic and since it has just been developed, there may be a number of issues.

And you may experience some buffering when going from file to live. This is unavoidable due to the nature of the function and low latency. If this is inconvenient, buffering issues can disappear if you add a little delay in advance by setting PartHoldBack in LLHLS to 5 or more. It is a choice between delay and buffering.

**Reply by nums** [reply]
Just tested with RTMP stream, it is working very well without adding latency compared to origin stream.
I will test tomorrow with SRT and multiple audio tracks.

Many thanks for this feature !

**Reply by getroot** [reply]
ScheduledChannel does not yet support multi audio tracks. To support multi audio tracks, mapping is difficult because all files and live inputs must always have the same number of audio tracks.

**Reply by nums** [reply]
Yes, I imagine the difficulty. If this could make things simpler: Should you specify in the XML file the number of audio tracks and return an error if the video or streaming does not have the same number of tracks?

**Reply by getroot** [reply]
Yes, that's how it should be implemented. Currently, if you configure Video only and there is no video in the item, an error is returned and a fallback is played.
However, although it is not that difficult to implement, it seems very difficult to use.

**Comment by basisbit** [comment]
Can those scheduled streams be consumed by all providers just like any normal incoming stream, for example for accessing it through OVT publisher for a Origin-Edge style configuration?

**Comment by getroot** [comment]
@basisbit 

The example below is a minimal example setup to transmit a file instead when the input is off-air using Scheduled Stream.

The stream set in this way is no different from the stream input from other providers, and can naturally be transmitted to OVT Publisher. Additionally, Scheduled Stream can use streams from any provider.

There is nothing different in Server.xml. Just configure the `<Schedule>` provider in `<Providers>` and configure the root directory of the media file and the root directory of the Schedule file. Schedule file is in `<Stream Name>.sch` format.

And you need to configure video/audio encoding. If set to bypass, the file and live profiles are different, so the video may be broken when switching.

### Server.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<Server version="8">
	<Name>OvenMediaEngine</Name>
	<Type>origin</Type>
	<IP>*</IP>
	<PrivacyProtection>false</PrivacyProtection>
	<StunServer>stun.ovenmediaengine.com:13478</StunServer>
	<Modules>
		<HTTP2>
			<Enable>true</Enable>
		</HTTP2>
		<LLHLS>
			<Enable>true</Enable>
		</LLHLS>
		<P2P>
			<Enable>false</Enable>
			<MaxClientPeersPerHostPeer>2</MaxClientPeersPerHostPeer>
		</P2P>
	</Modules>
	<Bind>
		<Providers>
			<RTMP>
				<Port>11935</Port>
				<WorkerCount>1</WorkerCount>
			</RTMP>
			<SRT>
				<Port>19998,19999</Port>
				<WorkerCount>1</WorkerCount>
			</SRT>
			<WebRTC>
				<Signalling>
					<Port>13333</Port>
					<TLSPort>13334</TLSPort>
					<WorkerCount>1</WorkerCount>
				</Signalling>
				<IceCandidates>
					<IceCandidate>192.168.0.160:13000/udp</IceCandidate>
					<TcpRelay>192.168.0.160:13478</TcpRelay>
					<TcpForce>true</TcpForce>
					<TcpRelayWorkerCount>1</TcpRelayWorkerCount>
				</IceCandidates>
			</WebRTC>
		</Providers>
		<Publishers>
            <OVT>
                <Port>19000</Port>
            </OVT>            
			<LLHLS>
				<Port>13333</Port>
				<TLSPort>13334</TLSPort>
				<WorkerCount>1</WorkerCount>
			</LLHLS>
			<WebRTC>
				<Signalling>
					<Port>13333</Port>
					<TLSPort>13334</TLSPort>
					<WorkerCount>1</WorkerCount>
				</Signalling>
				<IceCandidates>					
					<IceCandidate>192.168.0.160:12000/udp</IceCandidate>
					<TcpRelay>192.168.0.160:13478</TcpRelay>
					<TcpForce>false</TcpForce>
					<TcpRelayWorkerCount>1</TcpRelayWorkerCount>
				</IceCandidates>
			</WebRTC>
		</Publishers>
	</Bind>

	<VirtualHosts>
		<VirtualHost>
			<Name>default</Name>
			<Distribution>ovenServer</Distribution>
			<Host>
				<Names>
					<Name>ome-dev.airensoft.com</Name>
					<Name>*</Name>
				</Names>
				<TLS>
					<CertPath>../certificate.crt</CertPath>
					<KeyPath>../certificate.key</KeyPath>
					<ChainCertPath>../certificate.ca-bundle</ChainCertPath>
				</TLS>
			</Host>

			<Applications>
				<Application>
					<Name>app</Name>
					<Type>live</Type>

					<OutputProfiles>
						<OutputProfile>
							<Name>abr_stream</Name>
							<OutputStreamName>${OriginStreamName}</OutputStreamName>
							<Encodes>
								<Video>
									<Name>video_720</Name>
									<Codec>h264</Codec>
									<Profile>high</Profile>
									<Bitrate>1500000</Bitrate>
									<Framerate>30</Framerate>
									<Width>1280</Width>
									<Height>720</Height>
									<KeyFrameInterval>30</KeyFrameInterval>
									<Preset>medium</Preset>
									<BFrames>2</BFrames>
									<ThreadCount>4</ThreadCount>
								</Video>
								<Audio>
									<Name>aac_audio</Name>
									<Codec>aac</Codec>
									<Bitrate>128000</Bitrate>
									<Samplerate>48000</Samplerate>
									<Channel>2</Channel>
									<BypassIfMatch>
										<Codec>eq</Codec>
									</BypassIfMatch>
								</Audio>
							</Encodes>
						</OutputProfile>
					</OutputProfiles>
					<Providers>
						<SRT/>
						<RTMP/>
						<WebRTC/>
						<RTSPPull />

						<Schedule>
							<MediaRootDir>/data/getroot/scheduled</MediaRootDir>
							<ScheduleFilesDir>/data/getroot/scheduled</ScheduleFilesDir>
						</Schedule>
					</Providers>
					<Publishers>
						<AppWorkerCount>1</AppWorkerCount>
						<StreamWorkerCount>16</StreamWorkerCount>
						<OVT />
						<LLHLS>
							<OriginMode>true</OriginMode>
							<ServerTimeBasedSegmentNumbering>true</ServerTimeBasedSegmentNumbering>
							<ChunkDuration>1</ChunkDuration>
							<PartHoldBack>3</PartHoldBack>
							<EnablePreloadHint>true</EnablePreloadHint>
							<SegmentDuration>4</SegmentDuration>
							<SegmentCount>10</SegmentCount>
							<CrossDomains>
								<Url>*</Url>
							</CrossDomains>
						</LLHLS>
						<WebRTC>
							<Timeout>10000</Timeout>
							<Rtx>false</Rtx>
							<Ulpfec>false</Ulpfec>
							<JitterBuffer>false</JitterBuffer>
						</WebRTC>
					</Publishers>
				</Application>
			</Applications>
		</VirtualHost>
	</VirtualHosts>
</Server>
```

### stream.sch 

Create stream.sch in the folder you set and configure it as follows.

```
<?xml version="1.0"?>
<Schedule>
        <Stream>
                <Name>stream</Name>
                <BypassTranscoder>false</BypassTranscoder>
                <VideoTrack>true</VideoTrack>
                <AudioTrack>true</AudioTrack>
        </Stream>

        <FallbackProgram>
                <Item url="file://hevc.mov" start="0" duration="-1" />
                <Item url="file://avc.mov" start="0" duration="-1" />
        </FallbackProgram>

        <Program name="live" scheduled="2000-01-01T00:00:00.000+09" repeat="true">
                <Item url="stream://default/app/live" start="0" duration="-1" />
        </Program>
</Schedule>
```

- By default, the entered live broadcast is broadcast `<Program name="live">`. Check that it has been scheduled since 2000. This means that this program is unconditionally selected.
- When live input is off-air, transmit a pre-configured file `<FallbackProgram>`

**Comment by bchah** [comment]
Hi @getroot, this feature is amazing! And working better than ever with the latest improvements.

I have a feature request that would be quite useful: would you consider allowing a *directory* to be used as a program item? The idea being that we could point OME at a directory e.g. "cat_videos" and it would simply play through the items in the folder alphabetically, allowing dynamic addition / removal of new items without changing the stream.sch config.

**Reply by getroot** [reply]
Great idea. But handling it in OME can be quite complicated. How about having external software watch the directory and create/update the sch file?

**Reply by basisbit** [reply]
+1 on doing this in external software, hoping to avoid feature-creep

**Reply by bchah** [reply]
I created PR #1503 with a shell script that fulfills this need for now. It requires inotify-tools and xmlstarlet. Hopefully it is useful and inspires you to put this functionality into the application 🙏

**Comment by Manfred** [comment]
I tried creating a persistent live channel exactly as suggested in a  [previous comment](https://github.com/AirenSoft/OvenMediaEngine/discussions/1461#discussioncomment-8059664) and I found that it works great when I use OBS and RTMP as the source of the stream (meaning `stream://default/app/live`). However, when I try to use OBS + WHIP or browser based WebRTC as the stream source it crashes OME almost immediately after the stream is published with the following output:

```
[Stream Info]
id(100), msid(2), output(stream), SourceType(Scheduled), RepresentationType(Source), Created Time (Wed Oct 16 13:55:11 2024) UUID(f58864e9-4e6c-4df5-8085-50a9f8a345ec/default/#default#app/stream/i)

	Video Track #0: Public Name(Video_0) Variant Name(Video) Bitrate(0b) Codec(1,H264,default:0) BSF(H264_RTP_RFC_6184) Resolution(0x0) Framerate(0.00) KeyInterval(1/frame) SkipFrames(-1) BFrames(0) timebase(1/90000)
	Audio Track #1: Public Name(Audio_1) Variant Name(Audio) Bitrate(1.20Kb) Codec(8,OPUS,default) BSF(OPUS_RTP_RFC_7587) Samplerate(48.0K) Format(none, 0) Channel(stereo, 2) timebase(1/48000)
	Data  Track #2: Public Name() Variant Name(Data) Codec(0,Unknown,none) BSF(ID3v2) timebase(1/1000)
[2024-10-16 13:55:21.046] C [Scheduled:66] OvenMediaEngine | signals.cpp:148  | OME v0.17.1 ((From archive)) received signal 11 (SIGSEGV), interrupt.
```

**Reply by getroot** [reply]
Thanks for reporting the issue. Could you please submit this as a bug report to this issue?

**Comment by getroot** [comment]
Over time, the Scheduled Channel has undergone numerous stability updates. We have conducted extensive testing in real-world production environments, and we now believe everything is running perfectly.

For optimal performance when using the ScheduledChannel, it is recommended to enable Video/Audio encoding. We have extensively researched seamless source content transitions, and the encoder in OvenMediaEngine handles various tasks, such as timestamp correction during source content switching and audio data adjustments for A/V synchronization.

Another great feature is that the Scheduled Channel now supports multi-audio track input. Simply define the AudioMap for the Stream in the Schedule file as shown below. One important note: all source content must have at least as many audio tracks as defined in the AudioMap.

```
<?xml version="1.0"?>
<Schedule>
	<Stream>
		<Name>today</Name>
		<BypassTranscoder>false</BypassTranscoder>
		<VideoTrack>true</VideoTrack>
		<AudioTrack>true</AudioTrack>
		<AudioMap>
			<Item>
				<Name>English</Name>
				<Language>en</Language>
				<Characteristics>public.accessibility.describes-video</Characteristics>
			</Item>
			<Item>
				<Name>Korean</Name>
				<Language>ko</Language> 
				<Characteristics>public.alternate</Characteristics>
			</Item>
			<Item>
				<Name>Japanese</Name>
				<Language>ja</Language> 
				<Characteristics>public.alternate</Characteristics>
			</Item>
		</AudioMap>
	</Stream>
```

**Reply by steveRW** [reply]
Can the scheduled channel be used for live streams as well (not playing video files)?  Can it be used to switch seamlessly between multiple webrtc inputs?  Or is it only for scheduling pre-recorded videos?  Does OME have a way to do switching between multiple incoming lives streams to a single output (rtmp)?

**Comment by Manfred** [comment]
@steveRW I'm going to try to answer from my experience with OME from last year

> Can the scheduled channel be used for live streams as well (not playing video files)?

Yes, you can have both stream and prerecorded URLs as items in a program.

> Can it be used to switch seamlessly between multiple webrtc inputs?

Partially, the schedule means the server will play them items in the predefined sequence. So when you close the stream for item 1 it will go to item 2, etc.

There is also a noticeable switch over going from one item to the next. Sort of like when you switch cameras on desktop streaming software.

> Does OME have a way to do switching between multiple incoming lives streams to a single output (rtmp)?

Sort of, you can have two streams in a schedule and connect and disconnect them using a stateful frontend. Let's say you have the following schedule:

* Stream 1
* Stream 2

You start with stream 1 connected, connect stream 2, close stream 1, now you are on stream 2. When the program is configured to repeat you can do the reverse to switch back to stream 1.

**Reply by getroot** [reply]
For additional explanation, it is good to configure it as `<FallbackProgram>` for redundancy.

And the schedule file (`.sch`) can be changed dynamically. Of course, it can be done with the API. Therefore, it is also possible to switch the input stream by modifying the .sch file.

**Comment by techno-club** [comment]
Great feature! I've been using shell scripts that call the ffmpeg cli that are executed by the at" daemon for my current playlist/channel functionality.  

Can't wait to see if I can implement OME's Scheduled Channel for my use case.

Is it possible to have a series of videos like 10-12 playing on repeat 24/7  - that can be interrupted by a live input (rtmp) and when the live input has finished , the program continues to play (preferably) , or starts over.

**Reply by getroot** [reply]
The .sch file can be modified on the fly, so you can change your programs on the fly.

If your VoDs are fixed schedules and you don't want to change the .sch, schedule your live inputs as Programs. Then schedule your VoDs as FallbackPrograms. When there is no Live, the FallbackProgram will be played, and when there is Live input, the Program with the Live input will be played.

**Reply by techno-club** [reply]
So every Saturday, we play about 10-12 videos on loop  and that plays until Friday.  We have a mixture of live and pre-recorded on Friday and the again on Saturday, we have a new set of videos.   So is there a way to change the "Fallback Program" schedule each week?  That's the part that I'm missing.

**Reply by getroot** [reply]
We provide an API so that such business logic can be applied using external programs.
Alternatively, you can create a schedule file (.sch) several months in advance.
